### PR TITLE
User warnings large upload

### PR DESCRIPTION
**Pull request by hh-space-invader**
Add user warnings if uploading or loading collection with more than 20K points in local mode.
I've refactored the way we do DeprecationWarning as we also want to add  UserWarning to be shown once by adding  show_warning and show_warning_once and then specify the category of the warning needed to be shown once.

### All Submissions:

* [ ] Contributions should target the `dev` branch. Did you create your branch from `dev`?
* [ ] Have you followed the guidelines in our Contributing document?
* [ ] Have you checked to ensure there aren't other open [Pull Requests](../../../pulls) for the same update/change?

### New Feature Submissions:

1. [ ] Does your submission pass tests?
2. [ ] Have you installed `pre-commit` with `pip3 install pre-commit` and set up hooks with `pre-commit install`?

### Changes to Core Features:

* [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your core changes, as applicable?
* [ ] Have you successfully ran tests with your changes locally?

### CODE REVIEW

File: qdrant_client/common/deprecations.py [hunk 1]
```diff
@@ -8,7 +8,7 @@ def deprecation_warning(message: str) -> None:
     warnings.warn(message, DeprecationWarning, stacklevel=2)
 
 
-def deprecation_warning_once(message: str, idx: Optional[str]) -> None:
+def deprecation_warning_once(message: str, idx: Optional[str] = None) -> None:
```
Comment (joein) [inline qdrant_client/common/deprecations.py:11]: why?

File: qdrant_client/local/local_collection.py [hunk 2]
```diff
@@ -140,6 +143,10 @@ def __init__(
         self.storage = None
         self.config = config
         if location is not None:
+            deprecation_warning_once(
```
Comment (joein) [inline qdrant_client/local/local_collection.py:146]: it's not a `deprecation` warning

File: qdrant_client/local/local_collection.py [hunk 3]
```diff
@@ -140,6 +143,10 @@ def __init__(
...
+                f"Local mode is not recommended for collections with more than {self.LARGE_DATA_THRESHOLD:,} points. Consider using Qdrant server mode for better performance with large datasets.",
```
Comment (joein) [inline qdrant_client/local/local_collection.py:147]: It's probably unclear what is `server` mode, let's write about hosting qdrant in docker or using cloud

File: qdrant_client/local/local_collection.py [hunk 4]
```diff
@@ -2163,10 +2170,20 @@ def _upsert_point(self, point: models.PointStruct) -> None:
...
+                )
             for point in points:
                 self._upsert_point(point)
         elif isinstance(points, models.Batch):
             batch = points
+            if len(batch.ids) > self.LARGE_DATA_THRESHOLD:
+                deprecation_warning_once(
+                    f"Local mode is not recommended for collections with more than {self.LARGE_DATA_THRESHOLD:,} points. Consider using Qdrant server mode for better performance with large datasets.",
+                    idx="large-data-local",
+                )
```
Comment (joein) [inline qdrant_client/local/local_collection.py:2186]: it's not about the amount of points they're trying to upsert with one call, it's about the amount of points in the collection in general

**Review by joein — CHANGES_REQUESTED** [review]

**Review by joein — CHANGES_REQUESTED** [review]

File: qdrant_client/common/client_warnings.py [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+    message: str, category: type[Warning] = UserWarning, idx: Optional[str] = None
+) -> None:
+    """
+    Show a warning of the specified category only once per program run.
+    """
+    key = idx if idx is not None else message
+
+    if key not in SEEN_MESSAGES:
+        SEEN_MESSAGES.add(key)
+        show_warning(message, category)
```
Comment (generall) [inline qdrant_client/common/client_warnings.py:21]:
```suggestion
        show_warning(message, category)
        
def deprecation_warning(message: str) -> None:
    show_warning(message, DeprecationWarning)


def deprecation_warning_once(message: str, idx: Optional[str]) -> None:
    show_warning_once(message, DeprecationWarning, idx)
```

Changes of unrelated methods could be avoided if you just add those

File: qdrant_client/local/local_collection.py [hunk 6]
```diff
@@ -2158,6 +2161,20 @@ def _upsert_point(self, point: models.PointStruct) -> None:
             self.storage.persist(point)
 
     def upsert(self, points: Union[Sequence[models.PointStruct], models.Batch]) -> None:
+        points_count = (
+            len(points)
+            if isinstance(points, list)
+            else len(points.ids)
+            if isinstance(points, models.Batch)
+            else 0
+        )
```
[the remark this answers is not in the document]
Comment (joein) [reply qdrant_client/local/local_collection.py:2170]: I would even suggest removing this check and just check collection size in the end of the method
1) It is easier just to check len(self.ids) in the end
2) upsert can contain points with the same ids which will not lead to exceeding `self.LARGE_DATA_THRESHOLD`

**Review by joein — CHANGES_REQUESTED** [review]

File: qdrant_client/local/async_qdrant_local.py [hunk 7]
```diff
@@ -98,11 +101,18 @@ def _load(self) -> None:
...
-                    self.collections[collection_name] = LocalCollection(
+                    collection = LocalCollection(
                         config,
                         collection_path,
                         force_disable_check_same_thread=self.force_disable_check_same_thread,
                     )
+                    self.collections[collection_name] = collection
+                    if len(collection.ids) > self.LARGE_DATA_THRESHOLD:
+                        show_warning_once(
+                            message=f"Local mode is not recommended for collections with more than {self.LARGE_DATA_THRESHOLD:,} points. Consider using Qdrant docker (http/grpc) or Qdrant cloud for better performance with large datasets.",
```
[the remark this answers is not in the document]
Comment (joein) [reply qdrant_client/local/async_qdrant_local.py:112]: same for the others

**Review by joein — APPROVED** [review]

**Pull request merged**